include/nbl/builtin/hlsl/sort/counting.hlsl

```diff
@@ -0,0 +1,137 @@
+// Copyright (C) 2018-2024 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+
+#include "nbl/builtin/hlsl/sort/common.hlsl"
+#include "nbl/builtin/hlsl/workgroup/arithmetic.hlsl"
+
+#ifndef _NBL_BUILTIN_HLSL_SORT_COUNTING_INCLUDED_
+#define _NBL_BUILTIN_HLSL_SORT_COUNTING_INCLUDED_
+
+namespace nbl
+{
+namespace hlsl
+{
+namespace sort
+{
+
+template<
+    uint16_t GroupSize,
+    uint16_t KeyBucketCount,
+    typename Key,
+    typename KeyAccessor,
+    typename ValueAccessor,
+    typename ScratchAccessor,
+    typename SharedAccessor,
+    bool robust=false
+>
+struct counting
+{
+    uint32_t inclusive_scan(uint32_t value, NBL_REF_ARG(SharedAccessor) sdata)
+    {
+        return workgroup::inclusive_scan < plus < uint32_t >, GroupSize >::
+                template __call <SharedAccessor>(value, sdata);
+    }
+
+    void build_histogram(NBL_REF_ARG( KeyAccessor) key, NBL_REF_ARG(SharedAccessor) sdata, const CountingParameters<Key> data)
+    {
+        uint32_t tid = workgroup::SubgroupContiguousIndex();
+        uint32_t buckets_per_thread = (KeyBucketCount + GroupSize - 1) / GroupSize;
+        uint32_t baseIndex = (glsl::gl_WorkGroupID().x * GroupSize) * data.elementsPerWT;
+
+        for (int i = 0; i < buckets_per_thread; i++) {
+            uint32_t prev_bucket_count = GroupSize * i;
+            sdata.set(prev_bucket_count + tid, 0);
+        }
+
+        sdata.workgroupExecutionAndMemoryBarrier();
+
+        for (int i = 0; i < data.elementsPerWT; i++)
+        {
+            uint32_t prev_element_count = GroupSize * i;
+            int j = baseIndex + prev_element_count + tid;
+            if (j >= data.dataElementCount)
+                break;
+            uint32_t k = key.get(j);
+            if (robust && (k<data.minimum || k>data.maximum) )
+                continue;
+            sdata.atomicAdd(k - data.minimum, (uint32_t) 1);
+        }
+
+        sdata.workgroupExecutionAndMemoryBarrier();
+    }
+
+    void histogram(NBL_REF_ARG( KeyAccessor) key, NBL_REF_ARG(ScratchAccessor) scratch, NBL_REF_ARG(SharedAccessor) sdata, const CountingParameters<Key> data)
+    {
+        uint32_t tid = workgroup::SubgroupContiguousIndex();
+        uint32_t buckets_per_thread = (KeyBucketCount + GroupSize - 1) / GroupSize;
```

Comment:
1. slightly more readable "round up division" is `(x-1)/d+1`
2. another pattern to do what you want here is
```cpp
for (tid<KeyBucketCount; tid+=GroupSize)
````
instead of
```cpp
for (int i=0; i<buckets_per_thread; i++)
```

Reply:
Done